ui/src/websocket/queryStream.ts

```diff
@@ -0,0 +1,177 @@
+  const { children } = props;
+  const [ publicTemplateIds, setPublicTemplateIds ] = useState<string[]>([]);
+  const [ partyTemplateIds, setPartyTemplateIds ] = useState<string[]>([]);
+
+  const [ partyToken, setPartyToken ] = useState<string>();
+  const [ publicToken, setPublicToken ] = useState<string>();
+
+  const [ streamErrors, setStreamErrors ] = useState<StreamErrors[]>();
+
+  useEffect(() => {
+    const token = retrieveCredentials()?.token;
+    setPartyToken(token);
+  }, []);
+
+  const publicParty = useDablParties().parties.publicParty;
+  useEffect(() => {
+    getPublicToken(publicParty).then(token => {
+      if (token) {
+        setPublicToken(token);
+      }
+    })
+  }, [publicParty]);
+
+  const { contracts: partyContracts, errors: partyStreamErrors } = useDamlStreamQuery(partyTemplateIds, partyToken);
+  const { contracts: publicContracts, errors: publicStreamErrors } = useDamlStreamQuery(publicTemplateIds, publicToken);
+
+  useEffect(() => {
+    if (partyStreamErrors) {
+      setStreamErrors(errors => errors?.concat([partyStreamErrors]) || [partyStreamErrors]);
+    }
+
+    if (publicStreamErrors) {
+      setStreamErrors(errors => errors?.concat([publicStreamErrors]) || [publicStreamErrors])
+    }
+  }, [partyStreamErrors, publicStreamErrors]);
+
+  const subscribeTemplate = (templateId: string, asPublic?: boolean) => {
+    if (asPublic) {
+      setPublicTemplateIds(templateIds => [...templateIds, templateId]);
+    } else {
+      setPartyTemplateIds(templateIds => [...templateIds, templateId]);
+    }
+  }
+
+  const [ queryStream, setQueryStream ] = useState<QueryStream<T>>({
+    publicTemplateIds: [],
+    publicContracts: [],
+    partyTemplateIds: [],
+    partyContracts: [],
+    streamErrors: [],
+    subscribeTemplate,
+  });
+
+  useEffect(() => {
+    if (!_.isEqual(queryStream.partyContracts, partyContracts)) {
+      setQueryStream(queryStream => ({
+        ...queryStream,
+        partyTemplateIds,
+        partyContracts
+      }))
+    }
+  }, [partyContracts, partyTemplateIds, queryStream.partyContracts]);
+
+  useEffect(() => {
+    if (!_.isEqual(queryStream.publicContracts, publicContracts)) {
+      setQueryStream(queryStream => ({
+        ...queryStream,
+        publicTemplateIds,
+        publicContracts
+      }))
+    }
+  }, [publicContracts, publicTemplateIds, queryStream.publicContracts]);
+
+  useEffect(() => {
+    if (!_.isEqual(queryStream.streamErrors, streamErrors)) {
+      setQueryStream(queryStream => ({
+        ...queryStream,
+        streamErrors,
+      }))
+    }
+  }, [streamErrors, queryStream.streamErrors]);
+
+  return React.createElement(QueryStreamContext.Provider, { value: queryStream }, children);
+}
+
+// export function useContractQuery
+export function useContractQuery<T extends object, K = unknown, I extends string = string>(template: Template<T,K,I>, asPublic?: boolean) {
```

Comment:
Would it be possible to have a second function called `useContractQueryAsPublic` that just calls this one passing `true` into `asPublic` or would that cause problems with `useEffect`?

Reply:
yep, I think that will work, but I'm not sure I see the benefit to having separate functions if the underlying implementation manages both of the public/private views anyway

Reply:
yeah - the idea was to match `useStreamQueryAsPublic` and not have to worry about importing/passing a const around, but it's really personal preference as to which which you find cleaner, so I'm fine with either.